File: sdk/swift/sdkgen/generator_test.go
package sdkgen_test

import (
	"bytes"
	"encoding/json"
	"errors"
	"os"
	"path/filepath"
	"testing"

	"github.com/valksor/naatre/generator"
	"github.com/valksor/naatre/sdk/swift/sdkgen"
)

func TestGenerateReproducesCheckedArtifacts(t *testing.T) {
	root := filepath.Join("..", "..", "..")
	model := readFixture(t, filepath.Join(root, "conformance/v1/generator-model.json"))
	reference := readFixture(t, filepath.Join(root, "conformance/v1/generator-output.json"))
	first, err := sdkgen.Generate(model, reference)
	if err != nil {
		t.Fatalf("Generate: %v", err)
	}
	second, err := sdkgen.Generate(model, reference)
	if err != nil {
		t.Fatalf("second Generate: %v", err)
	}
	if !bytes.Equal(first.Source, second.Source) || !bytes.Equal(first.Manifest, second.Manifest) {
		t.Fatal("generation is not deterministic")
	}
	if !bytes.Equal(first.Source, readFixture(t, filepath.Join("..", "Sources/NaatreGenerated/Operations.swift"))) {
		t.Fatal("generated Swift source drifted")
	}
	if !bytes.Contains(first.Source, []byte("public struct GetAccountVariables: Codable, Sendable")) ||
		!bytes.Contains(first.Source, []byte("public init(from decoder: Decoder) throws")) {
		t.Fatal("generated variables are not Codable")
	}
	if !bytes.Equal(first.Manifest, readFixture(t, filepath.Join("..", "Sources/NaatreGenerated/operations.json"))) {
		t.Fatal("generated Swift manifest drifted")
	}
}

func TestGenerateRejectsReferenceDriftAndUnmappedScalar(t *testing.T) {
	root := filepath.Join("..", "..", "..")
	model := readFixture(t, filepath.Join(root, "conformance/v1/generator-model.json"))
	reference := readFixture(t, filepath.Join(root, "conformance/v1/generator-output.json"))
	drifted := bytes.Replace(reference, []byte("cc863005"), []byte("0c863005"), 1)
	_, err := sdkgen.Generate(model, drifted)
	var generated *sdkgen.Error
	if !errors.As(err, &generated) || generated.Code != "SWIFT_SDK_GENERATOR_REFERENCE_DRIFT" {
		t.Fatalf("error = %v, want SWIFT_SDK_GENERATOR_REFERENCE_DRIFT", err)
	}

	var value map[string]any
	if err := json.Unmarshal(model, &value); err != nil {
		t.Fatal(err)
	}
	configuration := value["configuration"].(map[string]any)
	delete(configuration["scalarMappings"].(map[string]any), "Money")
	changed, err := json.Marshal(value)
	if err != nil {
		t.Fatal(err)
	}
	_, err = sdkgen.Generate(changed, reference)
	generated = nil
	if !errors.As(err, &generated) || generated.Code != "SWIFT_SDK_GENERATOR_UNMAPPED_SCALAR" {
		t.Fatalf("error = %v, want SWIFT_SDK_GENERATOR_UNMAPPED_SCALAR", err)
	}
}

func TestGenerateRepresentsOpenEnumsAndUnionsWithoutReflection(t *testing.T) {
	root := filepath.Join("..", "..", "..")
	modelBytes := readFixture(t, filepath.Join(root, "conformance/v1/generator-model.json"))
	var value map[string]any
	if err := json.Unmarshal(modelBytes, &value); err != nil {
		t.Fatal(err)
	}
	schema := value["schema"].(map[string]any)
	types := schema["types"].([]any)
	admin := map[string]any{"id": "Admin", "name": "Admin", "kind": "object", "output": true, "fields": []any{map[string]any{"id": "Admin.id", "name": "id", "type": "ID", "required": true}}}
	search := map[string]any{"id": "SearchResult", "name": "SearchResult", "kind": "union", "output": true, "open": true, "variantMembers": []any{map[string]any{"id": "SearchResult.Admin", "type": "Admin"}, map[string]any{"id": "SearchResult.Account", "type": "Account"}}}
	schema["types"] = append(types, admin, search)
	changed, err := json.Marshal(value)
	if err != nil {
		t.Fatal(err)
	}
	reference, err := generator.Generate(changed)
	if err != nil {
		t.Fatal(err)
	}
	artifacts, err := sdkgen.Generate(changed, reference)
	if err != nil {
		t.Fatal(err)
	}
	if !bytes.Contains(artifacts.Source, []byte("public typealias SearchResult = OpenVariant")) ||
		!bytes.Contains(artifacts.Source, []byte("case unknown(String)")) {
		t.Fatalf("generated source does not preserve open variants:\n%s", artifacts.Source)
	}
}

func readFixture(t *testing.T, path string) []byte {
	contents, err := os.ReadFile(path)
	if err == nil {
		return contents
	}
	t.Fatalf("read fixture %q: %v", path, err)
	return nil
}

File: sdk/typescript/sdkgen/generator.test.mjs
import assert from "node:assert/strict";
import { createHash } from "node:crypto";
import { readFile } from "node:fs/promises";
import test from "node:test";

import { canonicalStringify } from "../runtime/index.mjs";
import { TypeScriptGeneratorError, generate, generatorVersion } from "./generator.mjs";

const modelURL = new URL("../../../conformance/v1/generator-model.json", import.meta.url);
const referenceURL = new URL("../../../conformance/v1/generator-output.json", import.meta.url);

test("generator reproduces checked artifacts byte for byte", async () => {
  const [model, reference, source, manifest] = await Promise.all([
    readFile(modelURL),
    readFile(referenceURL),
    readFile(new URL("../generated/operations.ts", import.meta.url), "utf8"),
    readFile(new URL("../generated/operations.json", import.meta.url), "utf8"),
  ]);
  const first = generate(model, reference);
  const second = generate(model, reference);
  assert.equal(first.source, source);
  assert.equal(first.manifest, manifest);
  assert.deepEqual(first, second);
  assert.match(first.source, new RegExp(generatorVersion));
  assert.match(first.source, /unknown: string/u);
});

test("generator rejects drift, unmapped scalars, and normalized symbol collisions", async () => {
  const [modelBytes, referenceBytes] = await Promise.all([readFile(modelURL), readFile(referenceURL)]);
  const reference = JSON.parse(referenceBytes);
  reference.operations[0].persisted.digest = "0".repeat(64);
  assert.throws(() => generate(modelBytes, JSON.stringify(reference)), generatorCode("TYPESCRIPT_SDK_GENERATOR_REFERENCE_DRIFT"));

  const model = JSON.parse(modelBytes);
  delete model.configuration.scalarMappings.Money;
  assert.throws(() => generate(JSON.stringify(model), referenceBytes), generatorCode("TYPESCRIPT_SDK_GENERATOR_UNMAPPED_SCALAR"));

  const collision = JSON.parse(modelBytes);
  collision.schema.types.push({ id: "status", name: "status", kind: "scalar", input: true, output: true });
  collision.configuration.scalarMappings.status = "string";
  assert.throws(() => generate(JSON.stringify(collision), referenceBytes), generatorCode("TYPESCRIPT_SDK_GENERATOR_SYMBOL_COLLISION"));

  const badMapping = JSON.parse(modelBytes);
  badMapping.configuration.scalarMappings.Money = "lossy-number";
  const badReference = JSON.parse(referenceBytes);
  badReference.scalarMappings.Money = "lossy-number";
  badReference.schema.digest = schemaDigest(badMapping.schema);
  assert.throws(() => generate(JSON.stringify(badMapping), JSON.stringify(badReference)), generatorCode("TYPESCRIPT_SDK_GENERATOR_UNMAPPED_SCALAR"));
});

test("generator represents open enums, tagged unions, input objects, one-of inputs, and deprecations", async () => {
  const [modelBytes, referenceBytes] = await Promise.all([readFile(modelURL), readFile(referenceURL)]);
  const model = JSON.parse(modelBytes);
  const reference = JSON.parse(referenceBytes);
  model.schema.types.push(
    { id: "Admin", name: "Admin", kind: "object", output: true, fields: [{ id: "Admin.id", name: "id", type: "ID", required: true }] },
    { id: "Choice", name: "Choice", kind: "oneof", input: true, fields: [{ id: "Choice.email", name: "email", type: "String", nullable: true }, { id: "Choice.phone", name: "phone", type: "String" }] },
    { id: "Filter", name: "Filter", kind: "input-object", input: true, description: "Filter input", fields: [{ id: "Filter.query", name: "query", type: "String", required: true, deprecation: { reason: "use term", replacement: "term" } }] },
    { id: "SearchResult", name: "SearchResult", kind: "union", output: true, open: true, variantMembers: [{ id: "SearchResult.Admin", type: "Admin" }, { id: "SearchResult.Account", type: "Account" }] },
    { id: "Tags", name: "Tags", kind: "list", input: true, output: true, element: "String", elementNullable: true },
    { id: "Translations", name: "Translations", kind: "map", input: true, output: true, element: "String" },
  );
  reference.schema.digest = schemaDigest(model.schema);
  const { source } = generate(JSON.stringify(model), JSON.stringify(reference));
  assert.match(source, /export type Status = "ACTIVE" \| "PENDING" \| \{ readonly unknown: string \};/u);
  assert.match(source, /readonly "\$type": "Account"; readonly "\$value": Account/u);
  assert.match(source, /readonly unknown: true/u);
  assert.match(source, /export interface Filter/u);
  assert.match(source, /@deprecated use term; use term/u);
  assert.match(source, /export type Choice = \{ readonly email: string \| null; readonly phone\?: never \} \| \{ readonly email\?: never; readonly phone: string \};/u);
  assert.match(source, /export type Tags = readonly \(string \| null\)\[\];/u);
  assert.match(source, /export type Translations = Readonly<Record<string, string>>;/u);
});

test("generated result decoding uses configured custom scalar codecs", async () => {
  const [modelBytes, referenceBytes] = await Promise.all([readFile(modelURL), readFile(referenceURL)]);
  const model = JSON.parse(modelBytes);
  const reference = JSON.parse(referenceBytes);
  model.operations[0].document.operations[0].select[0].$call.select[0].$field.name = "balance";
  model.operations[0].result.fields[0].result.fields[0].result.type = "Money";
  reference.operations[0].result.data = structuredClone(model.operations[0].result);
  const digest = semanticDigest("document", model.operations[0].document);
  reference.operations[0].persisted.digest = digest;
  reference.operations[0].request.persisted.digest = digest;
  const { source } = generate(JSON.stringify(model), JSON.stringify(reference));
  assert.match(source, /decodeScalar/u);
  assert.match(source, /decodeScalar\(codecs, "Money", value\)/u);
});

function generatorCode(code) {
  return (error) => error instanceof TypeScriptGeneratorError && error.code === code;
}

function schemaDigest(schema) {
  const value = structuredClone(schema);
  sortStrings(value, "capabilities");
  sortByID(value, "types");
  for (const descriptor of value.types) {
    for (const member of ["variants", "enumValues", "capabilities"]) sortStrings(descriptor, member);
    for (const member of ["fields", "enumMembers", "variantMembers", "retired", "traits"]) sortByID(descriptor, member);
    if (descriptor.entity) sortStrings(descriptor.entity, "keys");
    if (descriptor.scalar) sortStrings(descriptor.scalar, "acceptedWireShapes");
  }
  return semanticDigest("schema", value);
}

function semanticDigest(purpose, value) {
  return createHash("sha256").update(`naatre:${purpose}:c14n-1\n`).update(canonicalStringify(value)).digest("hex");
}

function sortStrings(value, member) {
  if (Array.isArray(value?.[member])) value[member].sort();
}

function sortByID(value, member) {
  if (Array.isArray(value?.[member])) value[member].sort((left, right) => left.id < right.id ? -1 : left.id > right.id ? 1 : 0);
}
